docs/proposals/20230612-node-os-image-caching.md

```diff
@@ -0,0 +1,230 @@
+
+Blessed Node - Whichever node has been the most stable and is chosen off of that metric as the one to clone the OS image of and cache it for future nodes to use
+
+## Summary
+
+A controller will be added to CAPZ which caches the Nodes’ OS image on a regular interval and updates the model to use that image for future scale outs. This feature will allow for:
+- faster horizontal scaling 
+- prevent security alerts from new nodes which immediately need a security update as those new nodes will now come with the security update already installed
+- help prevent users from needing to spin up warm nodes and overprovision
+
+## Motivation
+
+A model scenario would be an operator spinning up a CAPZ cluster and having this feature be able to be toggled on or off with an environment variable. If it was toggled on, then as the months passed and more security updates and patches needed to be applied to the operator’s node OS image, these changes would be cached on a regular interval and the operator would no longer have to wait for these changes to apply on new node creations. As a result, users will have faster horizontal scaling and require fewer warm nodes and overprovisioning to avoid this problem, especially since the new nodes will have the container images of the applications it will run pre-cached so pods will run quicker when scheduled. This feature will also help users have better security compliance as new nodes will already be compliant instead of needing to patch.
+
+#### Credits
+
+A big thank you to Jack Francis, Michael Sinz, and Amr Hanafi for their work on the Kamino project that used a similar Prototype Pattern with AKSEngine which inspired this feature request: https://github.com/jackfrancis/kamino
+
+### Goals
+
+1. A solution using the Node Prototype pattern which cache’s nodes’ OS image on a regular interval and replaces it as a new ‘golden image’.
+1. Faster horizontal scale outs
+1. Prevent security breaking issues on node bootup from security updates being required immediately
+
+We will know we’ve succeeded when we can benchmark speed increases and successful image changes
+
+### Non-Goals/Future Work
+
+1. Extend the functionality to Windows nodes
+1. Optimization for efficiency and scalability
+1. A more complicated method of selecting a good candidate node such as incorporating manual forced prototype creation, which would perform better but take research to find the optimal method. 
+1. Optimization of the default time interval into a specific, best general default which would have to be researched
+
+## Proposal
+
+### User Stories
+
+#### As an operator of CAPZ, I want faster horizontal scaling on new node deployment
+
+People setting up CAPZ clusters will be able to have faster horizontal scaling because their Nodes’s OS images will be continually cached, and thus they will be able to avoid longer pull times and avoid having to create warm nodes and overprovision.
```

Comment:
user stories should be in first person

Reply:
I think this user story sounds like it's more about reducing cost (both $$ and environmental) and improve sustainability because it will allow operators to not keep extra "warm nodes" around. The speed of scaling itself doesn't change (it actually gets a bit slower if you don't have the VM pre-provisioned) but what gets faster is going from VM to node running my application (starting pods), that's story 3.